// adonis5-scheduler/build/src/Scheduler/Locker.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const proper_lockfile_1 = __importDefault(require("proper-lockfile"));
const path_1 = __importDefault(require("path"));
const fs_1 = __importDefault(require("fs"));
class Locker {
    constructor(name, lockDir) {
        this.name = name;
        this.lockDir = lockDir;
        this.lockFilename = this._getLockFilePath();
        fs_1.default.mkdirSync(this.lockDir, { recursive: true });
        // fs.writeFileSync(this.lockFilename, '')
    }
    _getLockFilePath() {
        return path_1.default.join(this.lockDir, `${this.name}`);
    }
    check() {
        return proper_lockfile_1.default.check(this.lockFilename, { realpath: false });
    }
    lock() {
        return proper_lockfile_1.default.lock(this.lockFilename, { realpath: false });
    }
    unlock() {
        return proper_lockfile_1.default.unlock(this.lockFilename, { realpath: false });
    }
}
exports.default = Locker;
