// adonis5-scheduler/build/src/Exceptions/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RuntimeException = void 0;
const generic_exceptions_1 = require("@adonisjs/generic-exceptions");
/**
 * Class to throw runtime exceptions
 *
 * @class RuntimeException
 * @constructor
 */
class RuntimeException extends generic_exceptions_1.RuntimeException {
    /**
     * This exception is raised when task an undefined schedule
     */
    static undefinedTaskSchedule(task) {
        return new this(`${task} is not defined on schedule value`, 500, 'E_INVALID_TASK_SCHEDULE');
    }
    /**
     * This exception is raised when task an undefined handle
     */
    static undefinedTaskHandle(task) {
        return new this(`${task} is not defined on handle value`, 500, 'E_INVALID_TASK_HANDLE');
    }
    /**
     * This exception is raised when task an undefined handle
     */
    static undefinedInstanceTask(task) {
        return new this(`${task} is not extend of class Task`, 500, 'E_INVALID_TASK_INSTANCE');
    }
    /**
     * This exception is raised when task an undefined handle
     */
    static notFoundTask(dir) {
        return new this(`Not found task dir ${dir}`, 500, 'E_INVALID_SCHEDULER_DIR');
    }
}
exports.RuntimeException = RuntimeException;
